src/Infrastructure/ServiceBasedPlugin.php

```diff
@@ -0,0 +1,459 @@
+		$this->service_container->put(
+			static::SERVICE_PREFIX . static::INJECTOR_ID,
+			$this->injector
+		);
+
+		$services = $this->get_service_classes();
+
+		if ( $this->enable_filters ) {
+			/**
+			 * Filter the default services that make up this plugin.
+			 *
+			 * This can be used to add services to the service container for
+			 * this plugin.
+			 *
+			 * @param array<string> $services Associative array of identifier =>
+			 *                                class mappings. The provided
+			 *                                classes need to implement the
+			 *                                Service interface.
+			 */
+			$services = \apply_filters(
+				static::HOOK_PREFIX . static::SERVICES_FILTER,
+				$services
+			);
+		}
+
+		foreach ( $services as $id => $class ) {
+			$id    = $this->maybe_resolve( $id );
+			$class = $this->maybe_resolve( $class );
+
+			// Allow the services to delay their registration.
+			if ( is_a( $class, Delayed::class, true ) ) {
+				$registration_action = $class::get_registration_action();
+
+				if ( did_action( $registration_action ) ) {
+					$this->register_service( $id, $class );
+
+					continue;
+				}
+
+				\add_action(
+					$class::get_registration_action(),
+					function () use ( $id, $class ) {
+						$this->register_service( $id, $class );
+					}
+				);
+
+				continue;
+			}
+
+			$this->register_service( $id, $class );
+		}
+	}
+
+	/**
+	 * Register a single service.
+	 *
+	 * @param string $id ID of the service to register.
+	 * @param string $class Class of the service to register.
+	 */
+	protected function register_service( $id, $class ) {
+		// Only instantiate services that are actually needed.
+		if ( is_a( $class, Conditional::class, true )
+			&& ! $class::is_needed() ) {
+			return;
+		}
+
+		$service = $this->instantiate_service( $class );
+
+		$this->service_container->put( $id, $service );
+
+		if ( $service instanceof Registerable ) {
+			$service->register();
+		}
+	}
+
+	/**
+	 * Get the service container that contains the services that make up the
+	 * plugin.
+	 *
+	 * @return ServiceContainer Service container of the plugin.
+	 */
+	public function get_container() {
+		return $this->service_container;
+	}
+
+	/**
+	 * Instantiate a single service.
+	 *
+	 * @param string $class Service class to instantiate.
+	 *
+	 * @throws InvalidService If the service could not be properly instantiated.
+	 *
+	 * @return Service Instantiated service.
+	 */
+	protected function instantiate_service( $class ) {
+		/*
+		 * If the service is not registerable, we default to lazily instantiated
+		 * services here for some basic optimization.
+		 *
+		 * The services will be properly instantiated once they are retrieved
+		 * from the service container.
+		 */
+		if ( ! is_a( $class, Registerable::class, true ) ) {
+			return new LazilyInstantiatedService(
+				function () use ( $class ) {
+					return $this->injector->make( $class );
+				}
+			);
+		}
```

Comment:
Using lazily-instantiated services means that there's a certain level of automated optimization going on, where things are only instantiated on-demand when actually needed. So instead of hard-coding all service performance considerations, we can just have them all be known to the system, and that will then figure out what the smallest possible amount of instantiation is for a given request.